timemachine/cpp/src/nonbonded_all_pairs.cu

```diff
@@ -182,9 +201,15 @@ void NonbondedAllPairs<RealType, Interpolated>::hilbert_sort(
     const double *d_coords, const double *d_box, cudaStream_t stream) {
 
     const int tpb = 32;
-    const int B = (N_ + tpb - 1) / tpb;
+    const int B = (K_ + tpb - 1) / tpb;
 
-    k_coords_to_kv<<<B, tpb, 0, stream>>>(N_, d_coords, d_box, d_bin_to_idx_, d_sort_keys_in_, d_sort_vals_in_);
+    if (d_atom_idxs_) {
+        k_coords_to_kv_gather<<<B, tpb, 0, stream>>>(
+            K_, d_atom_idxs_, d_coords, d_box, d_bin_to_idx_, d_sort_keys_in_, d_sort_vals_in_);
+    } else {
+        // N_ == K_
+        k_coords_to_kv<<<B, tpb, 0, stream>>>(K_, d_coords, d_box, d_bin_to_idx_, d_sort_keys_in_, d_sort_vals_in_);
+    }
```

Comment:
We might be able to eliminate the conditional branch here by setting `d_atom_idxs_` to the range `1..N` and always using `k_coords_to_kv_gather` (will want to benchmark this)

Reply:
Thinking to leave this as-is for now to avoid any performance regression in the K=N case. Added a note to #639 to revisit this as part of clean / DRY tasks.